cura/Machines/MaterialManager.py

```diff
@@ -0,0 +1,324 @@
+from collections import defaultdict, OrderedDict
+from typing import Optional
+
+from PyQt5.Qt import QTimer, QObject, pyqtSignal
+
+from UM.Logger import Logger
+from UM.Settings import ContainerRegistry
+
+from cura.Machines.ContainerNode import ContainerNode
+
+
+class MaterialGroup:
+    __slots__ = ("name", "root_material_node", "derived_material_node_list")
+
+    def __init__(self, name: str):
+        self.name = name
+        self.root_material_node = None
+        self.derived_material_node_list = []
+
+    def __str__(self) -> str:
+        return "%s[%s]" % (self.__class__.__name__, self.name)
+
+
+class MaterialNode(ContainerNode):
+    __slots__ = ("material_map", "children_map")
+
+    def __init__(self, metadata: Optional[dict] = None):
```

Comment:
Wait, I thought that metadata was required?

Reply:
We will document this. The confusing part is that a `ContainerNode` here (which is a `MaterialNode`) is used as node in a tree structure. In this case, a node may or may contain a material container, so the metadata is optional here.

Reply:
Added documentation in c5d443109